lua/plenary/async_lib/async.lua

```diff
@@ -0,0 +1,158 @@
+local co = coroutine
+local uv = vim.loop
+
+local M = {}
+
+--- WIP idle stuff
+local thread_loop = function(thread, callback)
+  local idle = uv.new_idle()
+  idle:start(function()
+    local success = co.resume(thread)
+    assert(success, "Coroutine failed")
+
+    if co.status(thread) == "dead" then
+      idle:stop()
+      callback()
+    end
+  end)
+end
+
+-- use with wrap
+local execute = function(future, callback)
+  assert(type(future) == "function", "type error :: expected func")
+  local thread = co.create(future)
+
+  local step
+  step = function(...)
+    local res = {co.resume(thread, ...)}
+    local stat = res[1]
+    local ret = {select(2, unpack(res))}
```

Comment:
Can this be simplified to `local res, ret = co.resume(thread, ...)`?

Reply:
no, then multiple returns do not work

Reply:
Are you sure? I use multiple returns in multiple places in gitsigns and this is the line I have in my async lib.

Reply:
```lua
local co = coroutine

local func = function()
  return 1, 2, 3
end

local thread = coroutine.create(func)
local stat, ret = co.resume(thread)
print(stat)
print(ret) -- prints 1, ret only captures first return value
```

Reply:
Ah ok, thanks for checking. I guess gitsigns works because all returns are done via invoking callback functions as apposed to return.